src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs

```diff
@@ -0,0 +1,424 @@
+
+                libssl.SSL_free(sslPtr);
+            }         
+        }
+
+        //TODO should we check Bio should retry?
+        private static int BioRead(IntPtr BioPtr, IntPtr buffer, int count)
+        {
+            int bytes = libssl.BIO_read(BioPtr, buffer, count);
+            if (bytes != count)
+            {
+                throw CreateSslException("Failed in Read BIO");
+            }
+            return bytes;
+        }
+
+        //TODO should we check Bio should retry?
+        private static int BioWrite(IntPtr BioPtr, IntPtr buffer, int count)
+        {
+            int bytes = libssl.BIO_write(BioPtr, buffer, count);
+            if (bytes != count)
+            {
+                throw CreateSslException("Failed in Write BIO");
+            }
+            return bytes;
+        }
+
+        private static int GetSslError(IntPtr sslPtr, int result, string message)
+        {
+            int retVal = libssl.SSL_get_error(sslPtr, result);
+            if (retVal == libssl.SslErrorCode.SSL_ERROR_SYSCALL)
+            {
+                retVal = (int)libssl.ERR_get_error();
+            }
+            return retVal;
+        }
+
+        private static void SetSslCertificate(IntPtr contextPtr, IntPtr certPtr, IntPtr keyPtr)
+        {
+            int retVal = libssl.SSL_CTX_use_certificate(contextPtr, certPtr);
+            if (1 != retVal)
+            {
+                throw CreateSslException("Failed to use SSL certificate");
+            }
+
+            retVal = libssl.SSL_CTX_use_PrivateKey(contextPtr, keyPtr);
+            if (1 != retVal)
+            {
+                throw CreateSslException("Failed to use SSL certificate private key");
+            }
+            //check private key
+            retVal = libssl.SSL_CTX_check_private_key(contextPtr);
+            if (1 != retVal)
+            {
+                throw CreateSslException("Certificate pivate key check failed");
+            }
+
+        }
+
+        private static SslException CreateSslException(string message)
+        {
+            return new SslException(message);
+        }
+
+        private static SslException CreateSslException(string message, int error)
+        {
+            if (error == libssl.SslErrorCode.SSL_ERROR_SYSCALL)
+            {
+                return new SslException(message);
+            }
+            else
+            {
+                return new SslException(message, error);
+            }
+        }
+
+        private static SslException CreateSslException(IntPtr sslPtr, string message, int error)
+        {
+            return CreateSslException(message, libssl.SSL_get_error(sslPtr, error));
+        }
+
+        private sealed class SslException : Exception
+        {
+            public SslException(string message)
+                : base(message + ": " + Marshal.PtrToStringAnsi(libssl.ERR_reason_error_string(libssl.ERR_get_error())))
+            {
+                HResult = (int)libssl.ERR_get_error();
```

Comment:
Calling ERR_get_error changes the state of the error queue, but you've called it twice in this ctor.

It also would only apply if the SslErrorCode value was SSL_ERROR_SSL.